src/modules/rlm_ldap/user.c

```diff
@@ -35,6 +35,136 @@ USES_APPLE_DEPRECATED_API
+				"Enable sorting, or specify a more restrictive base_dn, filter or scope", cnt);
+			REDEBUG("The following entries were returned:");
+			RINDENT();
+			for (entry = ldap_first_entry(query->ldap_conn->handle, query->result);
+			     entry;
+			     entry = ldap_next_entry(query->ldap_conn->handle, entry)) {
+				dn = ldap_get_dn(query->ldap_conn->handle, entry);
+				REDEBUG("%s", dn);
+				ldap_memfree(dn);
+			}
+			REXDENT();
+			RETURN_MODULE_INVALID;
+		}
+	}
+
+	entry = ldap_first_entry(query->ldap_conn->handle, query->result);
+	if (!entry) {
+		ldap_get_option(query->ldap_conn->handle, LDAP_OPT_RESULT_CODE, &ldap_errno);
+		REDEBUG("Failed retrieving entry: %s",
+			ldap_err2string(ldap_errno));
+
+		RETURN_MODULE_FAIL;
+	}
+
+	dn = ldap_get_dn(query->ldap_conn->handle, entry);
+	if (!dn) {
+		ldap_get_option(query->ldap_conn->handle, LDAP_OPT_RESULT_CODE, &ldap_errno);
+		REDEBUG("Retrieving object DN from entry failed: %s", ldap_err2string(ldap_errno));
+
+		RETURN_MODULE_FAIL;
+	}
+	fr_ldap_util_normalise_dn(dn, dn);
+
+	RDEBUG2("User object found at DN \"%s\"", dn);
+
+	MEM(pair_update_control(&vp, attr_ldap_userdn) >= 0);
+	fr_pair_value_strdup(vp, dn, false);
+	ldap_memfree(dn);
+	if (user_ctx->out) *user_ctx->out = user_ctx->query;
+
+	return UNLANG_ACTION_CALCULATE_RESULT;
+}
+
+/** Initiate asynchronous retrieval of the DN of a user object
+ *
+ * Retrieves the DN of a user and adds it to the control list as LDAP-UserDN.
+ * Will also retrieve any attributes passed.
+ *
+ * This potentially allows for all authorization and authentication checks to be performed in one
+ * ldap search operation, which is a big bonus given the number of crappy, slow *cough*AD*cough*
+ * LDAP directory servers out there.
+ *
+ * @param[in] ctx	in which to allocate the query.
+ * @param[in] inst	rlm_ldap configuration.
+ * @param[in] request	Current request.
+ * @param[in] ttrunk	LDAP thread trunk to use.
+ * @param[in] attrs	Additional attributes to retrieve, may be NULL.
+ * @param[in] query_out	Where to put a pointer to the LDAP query structure -
+ *			for extracting extra returned attributes, may be NULL.
+ * @return
+ *	- UNLANG_ACTION_YIELD on success.
+ *	- UNLANG_ACTION_FAIL on failure.
+ */
+unlang_action_t rlm_ldap_find_user_async(TALLOC_CTX *ctx, rlm_ldap_t const *inst, request_t *request,
+					 fr_ldap_thread_trunk_t *ttrunk, char const *attrs[], fr_ldap_query_t **query_out)
+{
+	static char const	*tmp_attrs[] = { NULL };
+	ldap_user_find_ctx	*user_ctx;
+	LDAPControl		*serverctrls[] = { inst->userobj_sort_ctrl, NULL };
+
+	if (!attrs) memset(&attrs, 0, sizeof(tmp_attrs));
+
+	user_ctx = talloc_zero(ctx, ldap_user_find_ctx);
+	*user_ctx = (ldap_user_find_ctx) {
+		.inst = inst,
+		.ttrunk = ttrunk,
+		.attrs = attrs,
+		.out = query_out
+	};
+
+	if (inst->userobj_filter) {
+		if (tmpl_aexpand(user_ctx, &user_ctx->filter, request, inst->userobj_filter,
```

Comment:
Can you switch these out to the async variant if you're writing async code :)  I think it's somewhere in src/lib/unlang... It might be worth spending a little time figuring out how to flesh out that code for general consumption once this initial patch set is in.

Reply:
For yielding in between expansions it might be worth pushing an unlang function onto the stack, and then changing its resume function to represent the different expansion states.